Fix #947, #1483, #1247: allow inner type vars to be present in parent generic classes

daviskirk opened the pull request:
## Change Summary

Fix bugs where generic models with fields where the typevar is nested in another type `a: List[T]` are considered to be concrete. This allows these models to be subclassed as expected.

## Related issue number

Fixes #947, Fixes #1483, Fixes #1247
fix #2191
fix #2227

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

choogeboom commented on pydantic/generics.py:
```diff
@@ -113,4 +123,34 @@ def _parameterize_generic_field(field_type: Type[Any], typevars_map: Dict[TypeVa
 
 
 def _is_typevar(v: Any) -> bool:
```
I propose changing the name of this function to something better, since it's no longer doing a simple check to see if something is a type variable. Maybe something like `_is_generic` or `_has_typevar`?

daviskirk replied:
Good idea, I renamed it to `_has_typevar` as suggested

choogeboom reviewed (commented):
It's funny, I worked on this last Friday, but you beat me to the punch. I like your solution better than mine, but I think it's worth adding some test cases like the following:

https://github.com/wayfair-contribs/pydantic/blob/d2a1a5866283acf350338b1d33719f8b20598f5d/tests/test_generics.py#L204

https://github.com/wayfair-contribs/pydantic/blob/d2a1a5866283acf350338b1d33719f8b20598f5d/tests/test_generics.py#L220

daviskirk commented:
> It's funny, I worked on this last Friday, but you beat me to the punch. I like your solution better than mine, but I think it's worth adding some test cases like the following:
> 
> https://github.com/wayfair-contribs/pydantic/blob/d2a1a5866283acf350338b1d33719f8b20598f5d/tests/test_generics.py#L204
> 
> https://github.com/wayfair-contribs/pydantic/blob/d2a1a5866283acf350338b1d33719f8b20598f5d/tests/test_generics.py#L220

Well I'm glad I'm not alone on the generics code, there are some tricky problems and I'm really glad there are more eyes on this than just mine ;).  
I added the tests (as well as unifying the naming as I also realized there was already a test for subclasses just not testing the broken cases).  

However, I did change the "implicit parameters" test to not use implicit parameters as I think this would be more of a "new feature" (https://pydantic-docs.helpmanual.io/usage/models/#generic-models states that one has to inherit from typing.Generic in subclasses). 

I like the idea of implicit generic model inheritance but I think it should probably be discussed/handled in a separate pull request (I think your `_generate_class_parameters_from_bases` will probably work fine for that)

choogeboom reviewed (approved):
Changes look good! I agree that the implicit type hints is a new feature. I was delving into the mypy documentation on generics and forgot to cross reference the pydantic docs, so I missed the note about requiring all generic models to explicitly subclass `Generic`. I'll make an issue for it.

Not sure what the process is for getting a PR approved and merged, but #1913 mentions that @samuelcolvin is looking for people to review PRs. IMO this looks ready, and I hope it goes out quick!

choogeboom commented:
I created issue #2002 for the implicit generic parameters.

choogeboom commented:
Hmm... I found an edge case. Working to isolate it. I'll let you know once I figure it out.

choogeboom commented on pydantic/generics.py:
```diff
@@ -79,9 +84,14 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
+        concrete_type_args = tuple(resolve_type_hint(arg, typevars_map) for arg in type_args)
+        origin_type = _get_origin(type_) or type_
+        origin_type = _builtin_type_map.get(origin_type, origin_type)
+        return origin_type[concrete_type_args]
```
If `type_` is `Type[T]`, then `origin_type=type`. Unfortunately `type` is not subscriptable, so this will fail.
```suggestion
        if origin_type is type:
            return type_[concrete_type_args]
        else:
            return origin_type[concrete_type_args]
```
I'm not sure if this is a problem for any other types, but here's a test case you can use for `Type[T]`:

```
@skip_36
def test_generic_with_generic_type():
    T = TypeVar('T')

    class ModelWithType(GenericModel, Generic[T]):
        some_type: Type[T]

    class ReferenceModel(GenericModel, Generic[T]):
        abstract_base_with_type: ModelWithType[T]

    ReferenceModel[int]
```

daviskirk replied:
Thanks! Indeed there are other cases so my "_builtin_type_map" was way to specific.
I added the test and another variant.  python 3.9 will fix this with https://www.python.org/dev/peps/pep-0585/ but I worked around it for python < 3.9 / >= 3.6.

choogeboom reviewed (approved):
Nice

choogeboom commented on pydantic/generics.py:
```diff
@@ -79,9 +83,28 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
+
+    if lenient_issubclass(type_, GenericModel) and not type_.__concrete__:
+        return type_[tuple(resolve_type_hint(t, typevars_map) for t in type_.__parameters__)]
     return typevars_map.get(type_, type_)
```
Found another edge case, related to `Callable`. It turns out tat that the "input arguments" part of the type is a list of types, which means that `resolve_type_hints` fails. Here's a test case:

```py
@skip_36
def test_generic_with_callable():
    T = TypeVar('T')
    U = TypeVar('U')

    class VeryInnerModel(GenericModel, Generic[T, U]):
        t: T
        u: U
        some_callable: Callable[[Optional[int], str], None]

    VeryInnerModel[str, U]
```

This results in the following failure: `TypeError: unhashable type: 'list'` when `[Optional[int], str]` is used as a key in the `typevars_map`

choogeboom replied:
Turns out there are a number of problems with `Callable`:
1. Need to handle it in `resolve_type_hint` properly
2. Need to handle it in `_has_typevar` properly

choogeboom commented:
While I was investigating the `Callable` problem above, I came across another failing case

```
@skip_36
def test_nested_with_partial():
    T = TypeVar('T')
    U = TypeVar('U')

    class Model(GenericModel, Generic[T, U]):
        t: List[T]
        u: List[U]

    alias = Model[int, U]
    assert alias.__concrete__ is False
    concrete = alias[int]
    assert concrete.__concrete__ is True
```

Basically, what's happening is that we're not assigning `__parameters__` properly. We should be extracting all the parameters, but instead we're just passing the outer type, so for example in the above case, `alias` has `List[U]` as a parameter instead of `U`.

choogeboom commented on pydantic/generics.py:
```diff
@@ -79,9 +83,28 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
+        raise TypeError(f'Could not resolve type origin: {type_}')  # pragma: no cover
+
+    if lenient_issubclass(type_, GenericModel) and not type_.__concrete__:
+        return type_[tuple(resolve_type_hint(t, typevars_map) for t in type_.__parameters__)]
```
```suggestion
        return type_[tuple(resolve_type_hint(t, typevars_map) for t in type_.__parameters__)]
    if isinstance(type_, list):
        return [resolve_type_hints(element, typevars_map) for element in type_]        
```

choogeboom commented on pydantic/generics.py:
```diff
@@ -112,5 +135,35 @@ def _parameterize_generic_field(field_type: Type[Any], typevars_map: Dict[TypeVa
+def _has_typevar(v: Any) -> bool:
+    if lenient_issubclass(v, GenericModel):
+        return not v.__concrete__
+    args = _get_args(v)
```
```suggestion
    if isinstance(v, list):
        return any(_has_typevar(element) for element in v)
    args = _get_args(v)
```

choogeboom commented on pydantic/generics.py:
```diff
@@ -57,10 +61,10 @@ def __class_getitem__(cls: Type[GenericModelT], params: Union[Type[Any], Tuple[T
         created_model.__concrete__ = concrete
         if not concrete:
-            parameters = tuple(v for v in concrete_type_hints.values() if _is_typevar(v))
+            parameters = tuple(v for v in concrete_type_hints.values() if _has_typevar(v))
```
We need to extract the parameters from the concrete type hints rather than simply passing them along. Otherwise we end up with things like `Optional[T]` as a parameter.

choogeboom replied:
Nevermind. This is wrong.

choogeboom commented on pydantic/generics.py:
```diff
@@ -57,10 +61,10 @@ def __class_getitem__(cls: Type[GenericModelT], params: Union[Type[Any], Tuple[T
         created_model.__concrete__ = concrete
         if not concrete:
-            parameters = tuple(v for v in concrete_type_hints.values() if _is_typevar(v))
+            parameters = tuple(v for v in concrete_type_hints.values() if _has_typevar(v))
             parameters = tuple({k: None for k in parameters}.keys())  # get unique params while maintaining order
```
```suggestion
            parameters = tuple({v: None for v in typevars_map.values() if isinstance(v, TypeVar)})
```
Turns out that we were complicating this step more than necessary. We don't need `_has_typevar` at all.

choogeboom replied:
Argh... Turns out I was wrong, This fails if one of the params is generic.

choogeboom replied:
OK, I worked it out
```suggestion
            parameters = extract_parameters(typevars_map)
```
Where
```python
def extract_parameters(typevars_map: Dict[Any, Any]) -> Tuple[TypeVarType, ...]:
    extracted_parameters = {param: None for param in generate_parameters(list(typevars_map.values()))}
    return tuple(extracted_parameters)


def generate_parameters(v: Any) -> Iterator[TypeVarType]:
    if isinstance(v, TypeVar):
        yield v
    elif lenient_issubclass(v, GenericModel):
        yield from v.__parameters__
    elif isinstance(v, list):
        for var in v:
            yield from generate_parameters(var)
    else:
        args = _get_args(v)
        for arg in args:
            yield from generate_parameters(arg)
```

daviskirk commented:
Alright, great detective work!  I basically just copied your suggestion with small modification for mypy and linting.
I also removed _has_typevar and just used _generate_parameters everywhere since they were almost the same and added the tests: https://github.com/samuelcolvin/pydantic/pull/1989/commits/1b5d0f611f1992acc5d8e2722721faa67670068a

choogeboom commented on pydantic/generics.py:
```diff
@@ -61,12 +75,10 @@ def __class_getitem__(cls: Type[GenericModelT], params: Union[Type[Any], Tuple[T
-            parameters = tuple(v for v in concrete_type_hints.values() if _has_typevar(v))
-            parameters = tuple({k: None for k in parameters}.keys())  # get unique params while maintaining order
-            created_model.__parameters__ = parameters
+            created_model.__parameters__ = _extract_parameters(typevars_map)
```
Minor optimization:

```
new_params = _extract_parameters(typevars_map)
created_model.__concrete__ = bool(new_parameters)
created_model.__parameters__ = new_params
```

daviskirk replied:
correct, also removed extract parameters as it is now only used once and moved a test to it's counterpart that already existed

daviskirk commented:
rebased on master

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -99,15 +107,48 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
+            GenericAlias = typing.GenericAlias  # type: ignore
+        except AttributeError:
+            GenericAlias = ()  # for all python < 3.9
+        param_names = [
```
there's a lot going on in this comprehension. Might be more readable to pull it out into a boring old.

```py
param_names = []
for param in params:
    if ...:
        param_names.append(...)
```

samuelcolvin replied:
also please add comments explaining what's going on here.

daviskirk replied:
See below, got rid of the weird stuff with `pydantic.typing.display_as_type`

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -99,15 +107,48 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
         This method can be overridden to achieve a custom naming scheme for GenericModels
         """
-        param_names = [param.__name__ if hasattr(param, '__name__') else str(param) for param in params]
+        try:
```
is there a reason this needs to be in this function? Couldn't it be moved out to module level.

daviskirk replied:
I used `pydantic.typing.display_as_type` and added a case to that instead, which cleans this whole function up quite a lot.  Hope it's clearer now.

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -99,15 +107,48 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
-    if get_origin(type_) and getattr(type_, '__parameters__', None):
-        concrete_type_args = tuple([typevars_map[x] for x in type_.__parameters__])
-        return type_[concrete_type_args]
+def is_identity_typevars_map(typevars_map: Dict[TypeVarType, Type[Any]]) -> bool:
```
docstring please.

daviskirk replied:
I moved this to a more generic "all_identical" function (moved that to utils) and added some docs/examples/tests.  Hope it's clearer now.

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -99,15 +107,48 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
+    return bool(typevars_map) and all(k is v for k, v in typevars_map.items())
+
+
+def resolve_type_hint(type_: Any, typevars_map: Dict[Any, Any]) -> Any:
```
docstring please.

Also are you sure this doesn't overlap with any of the methods in `pydantic.typing`? Could it be moved into `pydantic.typing`? Or doe it need to be here.

daviskirk replied:
I looked through `pydantic.typing` again and there's nothing in there that does this.  

The code is quite geared towards the `generics` module (i.e. I don't see anywhere else in pydantic where this would be used), but the code itself is relatively generic in the sense that it just recursively replaces types according to a a map/dictionary.

It does have to reference the `GenericModel` though so I think I would keep it in there until there is somewhere else that needs this sort of recursive type map.

I also though there may have been confusion regarding the `resolve_annotations` in the `pydantic.typing` module, so I renamed this to `replace_types` since that's what it does (not really specific to type hints)

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -119,6 +160,20 @@ def check_parameters_count(cls: Type[GenericModel], parameters: Tuple[Any, ...])
 
 
+def iter_contained_typevars(v: Any) -> Iterator[TypeVarType]:
+    if isinstance(v, cast(type, TypeVar)):  # mypy < 0.790 needs this cast
```
we're now using `mypy` 0.790, so can this be removed?

Any docstring please.

daviskirk replied:
Better?  There's also a test specific to this function.  If you'd like I can add an example in the docstring, I'm unsure about the policy of docstring examples vs. tests (I don't think doctests are being run in ci if I'm not mistaken)

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -132,16 +187,16 @@ def _build_generic_fields(
 def _parameterize_generic_field(field_type: Type[Any], typevars_map: Dict[TypeVarType, Type[Any]]) -> Type[Any]:
-    if lenient_issubclass(field_type, GenericModel) and not field_type.__concrete__:
+    try:
+        is_model = isinstance(field_type, GenericModel) and not field_type.__concrete__
```
I'm not clear why this changed is required?

daviskirk replied:
Unfortunately `lenient_issubclass` fails if `field_type` is a generic alias.

```python
lenient_issubclass(list[int], GenericModel)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dek/Documents/Code/pydantic/pydantic/utils.py", line 145, in lenient_issubclass
    return isinstance(cls, type) and issubclass(cls, class_or_tuple)
  File "/home/dek/anaconda/envs/pydantic-39/lib/python3.9/abc.py", line 102, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
TypeError: issubclass() arg 1 must be a class
```

The exact logic is that:
```
isinstance(list[int], type)
```
returns `True` but then `issubclass` needs a `class` which `list[int]` is not.

I can add workaround code to lenient_issubclass but I wasn't sure that would be the right thing to do given how many places lenient_issubclass is used and this sort of handling of typehints seemed relatively specialized towards generics.

samuelcolvin commented on pydantic/typing.py:
```diff
@@ -80,8 +81,21 @@ def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:
+    else:
+        _GenericAlias = getattr(typing, '_GenericAlias')
+
+        def get_args(t: Type[Any]) -> Tuple[Any, ...]:
```
please rebase to master and check this is still required.

daviskirk replied:
Yes, this is already rebased.  Unfortunately, typing changes a lot in between 3.6 and 3.9.
For python `< 3.8`, we need to vendor `get_args` and do it correctly https://github.com/python/cpython/blob/4f3c25043d651a13c41cffcee703f7d5cb677cc7/Lib/typing.py#L1497 (mainly for all the generics stuff), but to do this correctly we need `_GenericAlias` which doesn't exist in python 3.6.
Since generics don't support 3.6 anyway that's ok for us (we just use the simpler version of `return getattr(t, '__args__', ())`, and for python >= 3.8 we can just use the cpython version. 
So unfortunately we need all three cases.

I'll add comments though explaining this

samuelcolvin commented on tests/test_generics.py:
```diff
@@ -1,14 +1,21 @@
 skip_36 = pytest.mark.skipif(sys.version_info < (3, 7), reason='generics only supported for python 3.7 and above')
 
+if sys.version_info >= (3, 9):
+    CompatList = list
```
this shouldn't be necessary any more.

daviskirk replied:
Very true, took it out, thanks!

samuelcolvin reviewed (commented):
I know pydantic isn't that well commented (and docstringed (word?)) at the moment. But for new complex stuff like this, let's try and add docstrings wherever possible.

I really don't understand all of generics off the top of my head and @dmontagu who wrote most of it isn't contributing to pydantic at the moment.

daviskirk commented:
> I know pydantic isn't that well commented (and docstringed (word?)) at the moment. But for new complex stuff like this, let's try and add docstrings wherever possible.
> 
> I really don't understand all of generics off the top of my head and @dmontagu who wrote most of it isn't contributing to pydantic at the moment.

I tried to update a few of the docstrings and added some comments to the more complicated steps of the "creation" part of the generics model.
Let me know if you want more or less.

samuelcolvin commented:
Sorry for the delay.

I think this is looking good, though I will admit that some of the nuance is lost on me.

Please could you fix coverage, otherwise happy to merge.

daviskirk commented:
> I think this is looking good, though I will admit that some of the nuance is lost on me.

Lots of the trickier parts are dealing with the changes (across python versions) and edge cases of the python typing module. If there are more issues like this that I don't run into myself or see in the bug tracker feel free to ping me now that I'm a bit familiar with the code... 
Unfortunately there will probably be stuff related to the generic aliases like (`list[int], `dict[str, int]` and so on) and in in 3.10 features (like the union operator `int | str`) that will also have some sort of edge cases that aren't handled yet (not complaining, both are really awesome features for writing code, but handling them in libraries like pydantic is still tricky). 

> Please could you fix coverage, otherwise happy to merge.

Fixed it, is back at 100% again.

samuelcolvin commented:
thank you so much @daviskirk. I think this has to be single most impressive contribution to pydantic in a long time.

ofek commented:
Will this be released soon?
